1. Training is available at these sites: [biostar Handbook and Penn State](https://www.biostarhandbook.com/); 
[UC Davis](http://bioinformatics.ucdavis.edu/training/). I am enrolled in the Penn State course. See the folder entitled [biostar handbook](https://github.com/shankar4/Functional-Genomics/tree/master/bioistar%20handbook) for my progress (and R code).  

2. Genomics Forum is [here](https://www.biostars.org/)

3. Databases: [Genbank at NCBI](https://www.ncbi.nlm.nih.gov/genbank/). **For more info, please see the folder entitled [Functl Gen Pevsnar](https://github.com/shankar4/Functional-Genomics/tree/master/FunctlGen%20Pevsner)**; 

4. Glossary (consolidated):

    * Alu - They are SINES (short interspersed elements) present in multiple copies in human and other primate genomes; about 300 bp in length, found in introns, 3' UTR, and intergenic genomic regions; mobile elements and has most abundant copy number (0.5 M). It is used as a universal PCR primer as it binds in both directions.
    * bacteriophase or phage - any virus that infects bacteria
    * domain - a discrete portion of a protein assumed to fold independently and has its own function
    * enhancer - enhance the rate of transcription; typically they are far away (1000s of bp away) from the sequence.
    * GeneID - unique ID in Entrez Gene; an integer which is species specific. Reported in RefSeq as a 'db_xref'.
    * homologous - similarity attributable to descent from a common ancestor
    * Map Viewer - available in Entrez Genomes. Can view the complete genome, display chromosome maps,and zoom in.
    * microsatellite - repititive stretches of short DNA sequences used as genetic markers to track inheritance in families. Also known as STRs (short tandem repets).
    * orthology - describes different species that derive from a common ancestor
    * paralog - a set of homologous genes that have diverged.
    * PDB - Protein Data Bank
    * promoter - A transcription factor attaches to this sequence to initiate transcription. Consists of 3 regions: Initiator box where transcription starts; TATA box 25 bases upstream from there; and 10 bp long upstream element - specific to a protein.
    * Pseudogene - defective duplicate copy of a genuine gene; cannot be expressed. 
    * Plasmid - circular molecule of double helical DNA; replicates independently of the bacterial cell's chromosome.
    * RNA Polymerase - binds to DNA during transcription and separates or unwinds the DNA strands. Three types and genes that they transcribe: I - genes for large ribosomal RNAs, II - genes that code for proteins, and III - genes for tRNA and other tiny RNAs.
    * QTL - Quantitative Trait Locus - a hypothesis that a certain small chromosomal region contributes significantly to a complex trait expression (based on density of linkage map)
    * recombination - merging of genes from two separate molecules of DNA
    * RefSeq - non-redundant reference sequences.
    * RepeatMaker - program that screens DNA sequence for interspersed repeats and low-complexity DNA sequences.
    * restriction - destruction of incoming foreign DNA by a bacterial cell
    * restriction enzyme - acts as molecular scissors to cut DNA into smaller manageable pieces.
    * splicing - cutting out the intronic regions. Telomere cap and poly-A tail are added prior to this.
    * transcription factors - Proteins that regulate RNA Polymerase II, by binding to certain specific DNA sequences,viz., promoter and enhancer
    * transduction - transport of genes from one cell to another with a virus particle as the transporter
    * Transformation - changing the properties of a bacterial cell as a result of uptake of pure DNA
    * UniGene cluster - sequences for known or putative genes. Complete dataset is here
    * WGS sequence - Whole Genome Shotgun sequence - uses large arrays of sequences to sequence rapidly the sequence.


    
   
    
   
    

