NCBI [reference](https://www.ncbi.nlm.nih.gov/snp/). Another good reference is [SNPedia](https://www.snpedia.com/index.php). People who have had genetic sequencing done pose questions and seek answers [here](https://www.reddit.com/r/SNPedia/). Wiki ref [here](https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism). Population codes are given [here](https://www.snpedia.com/index.php/Help_(population_diversity)). For eg., CEU is European, CHB is Han Chinese, GIH - Gujarathi Indians in Texas. Sample size is 90-180. Glossary of terms is [here](https://www.snpedia.com/index.php/Glossary)

After reading a paper recently entitled "Variation in FTO contributes to childhoood obesity and severe adult obesity" by Dina et al., 2007, I wished I could find a place with annotated summaries with cross-referencing. **SNPedia** does just that.  Meaning of rs Id from NCBI: A reference SNP ID number, or **“rs” ID**, is an identification tag assigned by NCBI to a group (or cluster) of SNPs that map to an identical location. The rs ID number, or rs tag, is assigned after submission (the db seems mature now; so no new tags). 

For e.g., from the releated page on [rs1121980](https://www.snpedia.com/index.php/Rs1121980): rs1121980, a SNP in the FTO gene, showed the strongest association of several SNPs in the region with early onset obesity in a study of ~1,000 Caucasians. The odds ratio for the rs1121980(T) risk allele was 1.66 (CI:1.3-2.01); and the odds ratios for heterozygotes and homozygotes (rs1121980(T;T)) were estimated (with a correct p<0.0001) at 1.67 (CI: 1.22-2.27) and 2.76....

Summarized from wiki: Single-nucleotide polymorphisms may be exonic, intronic, or intergenic. Exoniic SNPs do not necessarily change the amino acid sequence of the protein that is produced, due to degeneracy of the genetic code. Synonymous SNPs do not affect the protein sequence, while nonsynonymous SNPs change the amino acid sequence of protein. The nonsynonymous SNPs are of two types: **missense** (different codon coded) and **nonsense** (a stop codon introduced prematurely). 

Copied here a FAQ on other variants, from SNPedia:
SNPs only? what about CNVs, indels, inversions, epigenetics ... ?

dbSNP already assigns rs#s to small indels. A notable example is rs332. This is not a true single nucleotide polymorphism. Instead it is a deletion of three nucleotides with 2 different insertion texts. dbSNP handles it fine, and so does SNPedia. We also handle

    rs80358246 a 2500nt deletion
    rs17602729 multiallelics snp
    rs15793179 animal snps
    SL10030_490 plant snps
    i3003626 23andMe's internal identifiers
    gs100 haplotypes
    gs1064 haplogroup trees
    fully sequenced genomes (User:David Ewing Duncan, User:NA06985, User:NA06994, User:NA07357, ...)
    gs1013/criteria and genosets

dbVar now assigns names to CNVs, however we've not yet found any notable literature which uses these identifiers. As naming standards emerge for other types of variations SNPedia expects be able to handle them too 


