I donwoladed a few files to understand the formats and to explore ways to automate search and integrate functions. The .xml files are .kgml files that can be opened in a browser. The file 'ko00001.json' is the KEGG Orthology map. Open it in a browser. If you choose the 'raw' option, you will see all of the hierarchies; with the json option, you need to click and keep opening up the lower levels of hierarchies. The file 'br08901.json' is about KEGG networks. If you open it, you will find cortisol network, endocrine network, cortisol network, etc., with children nodes. Starting from the Brite menu [link](https://www.kegg.jp/kegg/brite.html), I chose 1. Genes and Proteins --> 1.4 Protein familes: signaling and cellular processes --> Proteoglycans. At this [page](https://www.kegg.jp/kegg-bin/get_htext?ko00535.keg), I clicked on 'Organism menu' link to find hsa (human ...). It took me to this [page](https://www.kegg.jp/kegg-bin/get_htext?hsa00535). On this page, it lists two main categories: cell surface PGs and ECM PGs. The downloaded json is 'hsa00535.json'. I see mention of aggrecan, versican, neurocan, brevican, etc here. It is a good hierarchy document. I then explored the diseases -> Cardiovascular diseases section. There are four subcategories: cardiac, vascular, hematologic, and hypertensive diseases. I found the following of interest to me:  H01626  Arteriosclerosis obliterans (H01626 - downloaded BRITE , under the 'vascular' category. But no mention of atherosclerosis or PAD.  I found Atherosclerosis under ICD-10 classification. Both mentions of arteriosclerosis to the same reference file: H06126, but different BR ref #s (br08402 and br08403) corresponding to the hierarchical files for humand diseases and human diseases in ICD-10 classification. Both are downloaded here. Under 'Human Diseases', Mental and behavioral disorders are under 'other diseases'. Diabetes is appropriately under 'Endocrine and metabolic diseases.' ICD-10 classification seems better and more 'standard.'  There is a file for new drug approvals in the US, for example. Obviously, this hierarchical document needs to be connected to the pathway document to integrate all the relevant objects for a given disease and all its molecular processes. I could not figure out whether that is feasible today. One can use R to integrate the information. Perhaps KEGGREST does. Yet to explore. 

Continuing..., Mismatch repair is under [Pathway Maps](https://www.kegg.jp/kegg/pathway.html): 2. Genetic Information Processing --> 2.4 Replication and repair --> 03430 Mismatch repair. Downloaded here. This is a highly conserved pathway that helps with maintaining genomic stability. Trying to relate... PMS2 on this [page](https://www.kegg.jp/kegg-bin/show_pathway?map=ko03430&show_description=show) yields K10858 if one hovers on that box. This has two entries in the ko03430.xml file. 

KEGG Modules is a collection of manually defined **functional** units (M #s). These are defined by a logical expression of K numbers (and other M numbers), allowing automatic evaluation of whether the gene set is complete, i.e. the module is present, in a given genome. Logical AND and OR operations are used. *To understand more on this*. Eg. M00002 represents glycolysis core module; hsa_M00002 the specific human pathway version. Note that though the map is manually drawn, modules are generated automatically. The specific module tracked here has a few boxes in green, and others in white, perhaps distinguishing what is available/missing in hsa. Not sure that is correct. I checked for different species, byy choosing a different one in a drop down box [here](https://www.kegg.jp/kegg-bin/show_module?map=M00002&org=myb&select_org=myb). Result is always the same. They have color code info [here](https://www.kegg.jp/kegg/kegg1c.html) - but that is not helpful. 
